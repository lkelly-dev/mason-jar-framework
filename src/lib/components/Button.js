import React, { useContext } from 'react';
import { cx, css } from 'emotion';
import MODIFIERS from './Modifiers';
import { debugAttrs, debugMode } from '../utils/debugging';

const ButtonOverride = css`
  -webkit-appearance: inherit !important;
  font-weight: bold;
  box-shadow: none !important;
`;

const Button = props => {
  const { context, children, color, outline, size, status, shape, className, as, ...rest } = props;
  const c =  (context) ? useContext(context) : null;
  const debug = debugMode(c)
  const buttonColor = MODIFIERS.COLOR[color] ? `btn-${color}` : null;
  const outlineColor = MODIFIERS.COLOR[outline] ? `btn-outline-${outline}` : null;
  const buttonSize = MODIFIERS.SIZE[size] ? `btn-${size}` : null;
  const buttonShape = MODIFIERS.SHAPE[shape] ? `btn-${shape}` : null;
  const buttonStatus = MODIFIERS.BUTTON[status] ? `${status}` : null;
  const Component = as || 'button';
  const buttonClass = cx(
    'btn',
    buttonColor,
    outlineColor,
    buttonSize,
    buttonShape,
    buttonStatus,
    ButtonOverride,
    className,
  );
  return (
    <Component type="button" className={buttonClass}
      {...(debug) ? debugAttrs(props): null}
      {...rest}>
      {children}
    </Component>
  );
};

const ButtonGroup = props => {
  const { context, children, size, orientation, className, ...rest } = props;
  const c =  (context) ? useContext(context) : null;
  const debug = debugMode(c)
  const buttonGroupSize = MODIFIERS.SIZE[size] ? `btn-group-${size}` : null;
  const buttonGroupOrientation = MODIFIERS.ORIENTATION[orientation] ? `btn-group-${orientation}` : null;
  const buttonGroupClass = cx('btn-group', buttonGroupSize, buttonGroupOrientation, className);
  return (
    <div className={buttonGroupClass}
      {...(debug) ? debugAttrs(props): null}
      {...rest}>
      {children}
    </div>
  );
};

Button.Group = ButtonGroup;

export { Button, ButtonGroup };
