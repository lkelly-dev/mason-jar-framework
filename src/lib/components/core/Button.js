import React from 'react';
import { cx, css } from 'emotion';
import MODIFIERS from './Modifiers';

const ButtonOverride = css`
  -webkit-appearance: inherit !important;
  font-weight: bold;
  box-shadow: none !important;
`;

const Button = props => {
  const { children, color, outline, size, status, shape, className, as, ...rest } = props;
  const buttonColor = MODIFIERS.COLOR[color] ? `btn-${color}` : null;
  const outlineColor = MODIFIERS.COLOR[outline] ? `btn-outline-${outline}` : null;
  const buttonSize = MODIFIERS.SIZE[size] ? `btn-${size}` : null;
  const buttonShape = MODIFIERS.SHAPE[shape] ? `btn-${shape}` : null;
  const buttonStatus = MODIFIERS.BUTTON[status] ? `${status}` : null;
  const Component = as || 'button';
  const buttonClass = cx(
    'btn',
    buttonColor,
    outlineColor,
    buttonSize,
    buttonShape,
    buttonStatus,
    ButtonOverride,
    className,
  );
  return (
    <Component type="button" className={buttonClass} {...rest}>
      {children}
    </Component>
  );
};

const ButtonGroup = props => {
  const { children, size, orientation, className, ...rest } = props;
  const buttonGroupSize = MODIFIERS.SIZE[size] ? `btn-group-${size}` : null;
  const buttonGroupOrientation = MODIFIERS.SIZE[orientation] ? `btn-group-${orientation}` : null;
  const buttonGroupClass = cx('btn-group', buttonGroupSize, buttonGroupOrientation, className);
  return (
    <div className={buttonGroupClass} {...rest}>
      {children}
    </div>
  );
};

Button.Group = ButtonGroup;

export { Button, ButtonGroup };
