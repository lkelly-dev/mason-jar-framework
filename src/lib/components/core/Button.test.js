import React from 'react';
import { shallow } from 'enzyme';
import { Button, ButtonGroup } from "./Button";

describe('Button', () => {
  it('matches snapshot', () => {
    const component = shallow(<Button></Button>);
    expect(component).toMatchSnapshot();
  })
  it('has expected props', ()=> {
    const component = shallow(
      <Button
        color="primary"
        outline="warning"
        size="sm"
        status="active"
        shape="rounded"
        className="nifty"
        whatever="dude">
        {"Cool"}
      </Button>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(classNameArray).toContain('btn');
    expect(classNameArray).toContain('btn-primary');
    expect(classNameArray).toContain('btn-outline-warning');
    expect(classNameArray).toContain('btn-sm');
    expect(classNameArray).toContain('btn-rounded');
    expect(classNameArray).toContain('active');
    expect(classNameArray).toContain('nifty');
    expect(classNameArray).toHaveLength(8);
    expect(component.prop('whatever')).toBe('dude');
  })
});

describe('ButtonGroup', () => {
  it('matches snapshot', () => {
    const component = shallow(<ButtonGroup></ButtonGroup>);
    expect(component).toMatchSnapshot();
  })

  it('has expected props', () => {
    const component = shallow(
      <ButtonGroup
        size="sm"
        orientation="vertical"
        className="neato"
        whatever="dude">
        <Button>wow</Button>
      </ButtonGroup>
    );
    const classNameArray = component.prop('className').split(' ');
    expect(component.contains(<Button>wow</Button>)).toEqual(true);
    expect(classNameArray).toContain('btn-group');
    expect(classNameArray).toContain('btn-group-sm');
    expect(classNameArray).toContain('btn-group-vertical');
    expect(classNameArray).toContain('neato');
    expect(component.prop('whatever')).toBe('dude');
  })
});

