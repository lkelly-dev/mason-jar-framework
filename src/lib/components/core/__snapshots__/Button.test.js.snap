// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Button matches snapshot 1`] = `
<button
  className="btn css-bh7dxm"
  type="button"
/>
`;

exports[`Button matches snapshot 1`] = `
<button
  className="btn css-bh7dxm"
  type="button"
/>
`;

exports[`ButtonGroup matches snapshot 1`] = `
<div
  className="btn-group"
/>
`;
